feat(http): Describe how reverse proxies transform ETags when content encoding changes (#35849)
mdn/content: files/en-us/web/http/conditional_requests/index.md

{{HTTPSidebar}}

HTTP has a concept of _conditional requests_, where the result, and even the success of a request, can be controlled by comparing the affected resources with a _validator_.

> These requests are useful for validating cached content, ensuring the integrity of a document when resuming a download, or preventing lost updates when uploading or modifying a document on the server.

Reviewer: FWIW, the most interesting point here is the first one. Perhaps something like:

```suggestion
These requests are useful for validating cached content, ensuring that it is only fetched if it differs from the copy that is already available to the browser.
Conditional requests are also useful for ensuring the integrity of a document when resuming a download, or preventing lost updates when uploading or modifying a document on the server.
```

Reviewer: Note, at this point I don't know how you might use an etag to ensure the integrity of a document etc. But if we make this statement here, ideally it should be expanded below.

Author: I guess if you have an ETag for some parts of versionA, but the file has changed to versionB since you've done a partial download, so you'll get different data if you resume.

Author: There's more details here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#integrity_of_a_partial_download

It is quite difficult to have a unique identifier for strong validation with {{HTTPHeader("Last-Modified")}}. Often this is done using an {{HTTPHeader("ETag")}} with the MD5 hash of the resource (or a derivative).

> [!NOTE]
> Because a change of content encoding requires a change to an ETag, some servers modify ETags when compressing responses from an origin server (reverse proxies, for example).

> > Apache Server appends the compression method (`-gzip`) to ETags by default, but this is [configurable using the `DeflateAlterETag` directive](https://httpd.apache.org/docs/2.4/mod/mod_deflate.html).

Reviewer 2: The "bug" or unexpected behavior is that they don't get removed when decompressing right?

Author: It's spec-compliant that two entities (uncompressed resource on the server and the one that's compressed at response-time) have distinct ETags so you can distinguish between them. You can test this:

```
curl -I https://example.com
HTTP/2 200
accept-ranges: bytes
age: 305826
cache-control: max-age=604800
content-type: text/html; charset=UTF-8
date: Fri, 13 Sep 2024 15:06:49 GMT
etag: "3147526947+gzip"
expires: Fri, 20 Sep 2024 15:06:49 GMT
last-modified: Thu, 17 Oct 2019 07:18:26 GMT
server: ECAcc (dcd/7D31)
x-cache: HIT
content-length: 1256

curl -I -H 'If-None-Match: "3147526947+gzip"' https://example.com
HTTP/2 304                  <- don't retransmit, the version the client has is fresh
accept-ranges: bytes
age: 181019
cache-control: max-age=604800
date: Fri, 13 Sep 2024 15:07:14 GMT
etag: "3147526947+gzip"
expires: Fri, 20 Sep 2024 15:07:14 GMT
last-modified: Thu, 17 Oct 2019 07:18:26 GMT
server: ECAcc (dcd/7D63)
x-cache: HIT
```

You need to use the exact ETag to get a cache hit & a 304.
There's some loss of efficiency with `GET` requests, but I can assume this is problematic if you're using unsafe methods. The Apache implementation changed a few times over the years, going from compressing the resource without changing the ETag, to converting the ETag to a 'weak' ETag, to now adding the suffix by default, but letting you configure behavior.

Author: The context here is that it might be unexpected that a reverse proxy (e.g., load balancer) in front of other servers is transforming requests / resources and changing the ETags before sending it further on to clients. With the 'weak' ETags, you would never get a match. Also [seen in Hapi](https://github.com/kanongil/hapi/commit/49c3945e5eb5308311e56747d3d68b52113659ff) etc.

> > Apache Server appends the compression method (`-gzip`) to ETags by default, but this is [configurable using the `DeflateAlterETag` directive](https://httpd.apache.org/docs/2.4/mod/mod_deflate.html).

Reviewer: Obvious later on, but not here. So I guess if a reverse proxy gets a request conditional for the uncompressed etag it would return the match if the compressed version hasn't changed too? 

```suggestion
> Apache Server appends the name of the compression method (`-gzip`) to ETags by default, but this is [configurable using the `DeflateAlterETag` directive](https://httpd.apache.org/docs/2.4/mod/mod_deflate.html).
```

Author: My understanding is: If the ETags don't match, the conditionals fail -> requesting uncompressed ETag (`abc`) when the proxy has `abc-gzip` cached means they're not equal and the proxy would make the request for the uncompressed version, compress it, then send `abc-gzip` as an ETag to the client. It's really unlikely that clients would ask for the non-suffixed version (how would they know about it?). Edge cases would be that clients would have a non-suffixed ETag (`abc`) before compression was turned on in the proxy, and this doesn't match on 1st conditional request when compression was enabled (bad timing, proxy config changed while a client was interacting with the server). 

After the first 200, the client would have `abc-gzip` of a compressed version and use that in subsequent requests (`304`).

Otherwise, some people might disable altering the ETags on compression, so there's no suffix, but that's not spec compliant.

### Weak validation

> Weak validation differs from strong validation, as it considers two versions of the document as identical if the content is equivalent. For example, a page that would differ from another only by a different date in its footer, or different advertising, would be considered _identical_ to the other with weak validation. These same two versions are considered _different_ when using strong validation. Building a system of ETags that creates weak validation may be complex, as it involves knowing the importance of the different elements of a page, but is very useful towards optimizing cache performance.

Reviewer: Wording of last sentence a bit odd.

```suggestion
Weak validation differs from strong validation, as it considers two versions of the document as identical if the content is equivalent. For example, a page that would differ from another only by a different date in its footer, or different advertising, would be considered _identical_ to the other with weak validation. These same two versions are considered _different_ when using strong validation. Building a system of ETags that uses weak validation is very useful for optimizing cache performance, but may be complex, as it involves knowing the importance of the different elements of a page.
```
